### PR TITLE
HHH-17738 : Add support for Oracle database AI Vector Search

### PR DESCRIPTION
**Pull request by Elab4d1**
https://hibernate.atlassian.net/browse/HHH-17738

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java [hunk 1]
```diff
@@ -142,7 +141,6 @@ public class SqlTypes {
 	 * as a synonym for {@link #VARCHAR}.
 	 *
 	 * @see org.hibernate.Length#LONG
-	 *
```
Comment (fmeheust) [inline hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java:145]: Should avoid making changed that are not related to your task.

File: hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java [hunk 2]
```diff
@@ -748,11 +748,24 @@ private StandardBasicTypes() {
 	 * The standard Hibernate type for mapping {@code float[]} to JDBC {@link org.hibernate.type.SqlTypes#VECTOR VECTOR},
 	 * specifically for embedding vectors like provided by the PostgreSQL extension pgvector.
 	 */
-	public static final BasicTypeReference<float[]> VECTOR = new BasicTypeReference<>(
-			"vector", float[].class, SqlTypes.VECTOR
+	public static final BasicTypeReference<double[]> VECTOR = new BasicTypeReference<>(
```
Comment (fmeheust) [inline hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java:751]: Why was it changed from float[] to double[]? If the change is necessary shouldn't the comment be updated ? Can it be different types for PostgreSQL and Oracle ?

File: hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java [hunk 3]
```diff
@@ -748,11 +748,24 @@ private StandardBasicTypes() {
...
+			"vector", double[].class, SqlTypes.VECTOR
+	);
+
+	public static final BasicTypeReference<byte[]> VECTOR_INT8 = new BasicTypeReference<>(
```
Comment (fmeheust) [inline hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java:755]: Add comment (see VECTOR). Save for VECTOR_FLOAT32 and VECTOR_FLOAT64

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorJdbcType.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
+package org.hibernate.vector;
```
Comment (fmeheust) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorJdbcType.java:1]: The classes OracleVectorJdbcType.java, OracleByteVectorJdbcType.java, OracleFloatVectorJdbcType.java and OracleDoubleVectorJdbcType.java share a big part of the code. Can this shared code be added to another class (maybe an abstract class, or a generic class) that will be extended by these classes? Just to avoid having several copies of the same lines of code

File: hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java [hunk 5]
```diff
@@ -649,31 +644,47 @@ public class SqlTypes {
...
 	public static final int NAMED_ENUM = 6001;
 
 
 	/**
 	 * A type code representing an {@code embedding vector} type for databases like
-	 * {@link org.hibernate.dialect.PostgreSQLDialect PostgreSQL} that have special extensions.
-	 * An embedding vector essentially is a {@code float[]} with a fixed size.
+	 * {@link org.hibernate.dialect.PostgreSQLDialect PostgreSQL} and {@link org.hibernate.dialect.OracleDialect Oracle 23ai}.
+	 * An embedding vector essentially is a fixed-size array of numeric values (float for PostgreSQL and double for Oracle)
 	 *
```
[the remark this answers is not in the document]
Comment (Michael-A-McMahon) [reply hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java:656]: Oracle can convert arrays of double, float, or byte to the VECTOR type. These primitive types correspond to the possible VECTOR types defined by Oracle Database: FLOAT64, FLOAT32, and INT8.

Oracle JDBC Doc: https://docs.oracle.com/en/database/oracle/oracle-database/23/jajdb/oracle/jdbc/OracleType.html#VECTOR

Oracle Database Doc: https://docs.oracle.com/en/database/oracle/oracle-database/23/vecse/create-tables-using-vector-data-type.html

But perhaps the intent is to keep things simple, and just support one Java type? If so, then double array is the right choice. A double can store the information of FLOAT64, FLOAT32, and INT8 values, without loss of precision.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+	private static final byte[] EMPTY = new byte[0];
+
+	public OracleByteVectorJdbcType(JdbcType elementJdbcType) {
+		super( elementJdbcType );
+	}
+
+	@Override
+	public <T> JdbcLiteralFormatter<T> getJdbcLiteralFormatter(JavaType<T> javaTypeDescriptor) {
+		return null;
```
Comment (Michael-A-McMahon) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:42]: Maybe just an FYI, but there is a literal format for VECTOR. The format is comma delimited numbers, within square brackets (same as what java.util.Arrays.toString(...) would return).

Example:
```sql
INSERT INTO example (vector_column) VALUES ('[0.1, 0.2, 0.3]')
```

Please note that I'm not familiar with Hibernate's code base; I'm not sure if JdbcLiteralFormatter is actually specified to generate a literal expression. The name implies it though, so I thought I should mention this.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+		return SqlTypes.VECTOR_INT8;
+	}
+
+	@Override
+	public String toString() {
+		return "OracleVectorTypeDescriptor";
+	}
+
+	private boolean isDriverBefore23(WrapperOptions options) throws SQLException {
+		final OracleConnection oracleConnection = options.getSession()
```
Comment (Michael-A-McMahon) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:68]: I'm thinking there could be a cost to acquiring a session. If we want to avoid that cost, we could instead use:
```java
DriverManager.getDriver("jdbc:oracle:thin:").getMajorVersion();
```

Comment (beikov) [reply hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:68]: You should be able to capture this information into `OracleServerConfiguration` and ultimately also in the `OracleDialect`, to then conditionally register respective `JdbcType` implementations.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	@Override
+	public <X> ValueBinder<X> getBinder(final JavaType<X> javaTypeDescriptor) {
+		return new BasicBinder<>( javaTypeDescriptor, this ) {
+			@Override
+			public void doBind(PreparedStatement st, X value, int index, WrapperOptions options) throws SQLException {
+				if ( isDriverBefore23( options ) ) {
+					st.setString( index, getStringVector( value, options ) );
+				}
+				else {
+					st.setObject( index, value, OracleType.valueOf( "VECTOR_INT8" ) );
```
Comment (Michael-A-McMahon) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:89]: I'm wondering if we could replace the `valueOf` call with `OracleType.VECTOR_INT8`? Or does this code need to compile with older versions of Oracle JDBC? If so, then I get it: OracleType.VECTOR_INT8 doesn't exist in versions before 23.4.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	@Override
+	public <T> JdbcLiteralFormatter<T> getJdbcLiteralFormatter(JavaType<T> javaTypeDescriptor) {
+		return null;
+	}
+
+	@Override
+	public void appendWriteExpression(String writeExpression, SqlAppender appender, Dialect dialect) {
+		appender.append( "vector(" );
```
Comment (Michael-A-McMahon) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:47]: I'm a bit puzzled by what this method is intended to do. It looks like it is generating DDL for a VECTOR column, but the DDL expression wouldn't have three arguments, it should only have length and type arguments. To clarify, I'm referring the the DDL syntax here: https://docs.oracle.com/en/database/oracle/oracle-database/23/sqlrf/Data-Types.html#GUID-801FFE49-217D-4012-9C55-66DAE1BA806F

I'm not familiar with Hibernate's code base, so maybe there's something obvious I'm not seeing. But if it's not obvious, I'd suggest adding a comment or JavaDoc to explain what this code is intending to do, and why it is correct to do it. This will help other programmers who might also be confused by it.

Comment (fmeheust) [reply hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:47]: I believe it is the VECTOR constructor: https://docs.oracle.com/en/database/oracle/oracle-database/23/vecse/vector-constructors.html#GUID-C589E48C-1D34-4E84-8312-222AA975F129

Comment (Michael-A-McMahon) [reply hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:47]: Makes sense, thank you!

Comment (Elab4d1) [reply hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:47]: Yes, it is the VECTOR contructor. I believe we can use "to_vector" contructor to avoid confusion between the DDL and the VECTOR contructor.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java (hunk 8 above)
Comment (Elab4d1) [reply hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:89]: You are right, these types only register if OracleDialect is detected, and the user may have an older version of Oracle JDBC, that's why I used `valueOf` method instead of `OracleType.VECTOR_INT8`

File: hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java (hunk 2 above)
Comment (Elab4d1) [reply hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java:751]: You are right, in the JDBC docs, they said "The preferred array class for SQL to Java conversions for VECTOR is double[], as a double can store the information of any VECTOR value without losing information." https://docs.oracle.com/en/database/oracle/oracle-database/23/jajdb/oracle/jdbc/OracleType.html#VECTOR, but since PostgreSQL only supports float values, we can either create basicTypeReference with different types for each db, or just use float for both. what do you think ?

Comment (Michael-A-McMahon) [reply hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java:751]: Is there existing code that relies on Hibernate to convert a SQL VECTOR to a float[]? We don't want to break that code by changing it to a double[].

I'd recommend leaving float[] as the default mapping. I'm sure the Hibernate team has already thought about why float[] is a sensible default, and they chose it for a good reason.

I'm not an ML programmer, but FWIW: I think 32-bit precision represents a sensible default for ML programming.

For Oracle Database, we do support a VECTOR of 64-bit floating point numbers. However, Oracle JDBC can convert that data into 32-bit floats. It will apply a narrowing conversion, as specified in the "SQL to Java Conversions" section here: http://ojdbc.us.oracle.com/scratch/JDBC_Javadoc/MAIN/latest/oracle/jdbc/OracleType.html#VECTOR_FLOAT64

BTW, I'm the one who wrote the OracleType JavaDocs. While I did say that double[] is _preferred_, it does not mean it is _required_.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/type/StandardBasicTypes.java:751]: We choose `float[]` as default because it's the datatype of choice for most ML models and pgvector also only supports float32 or smaller at this point. So yes, please revert this change.

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+		super( elementJdbcType );
+		this.javaType = javaType;
+	}
+
+	public abstract String getFriendlyName();
+
+	public abstract void appendWriteExpression(String writeExpression, SqlAppender appender, Dialect dialect);
+
+	@Override
+	public int getJdbcTypeCode() {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:56]: `getJdbcTypeCode()` is what we use for binding `null` and understanding the JDBC concept. What you want to change here is the SQL type code:

```suggestion
	public int getDefaultSqlTypeCode() {
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+	public String toString() {
+		return "OracleVectorTypeDescriptor";
+	}
+
+	/**
+	 * Checks if the Oracle JDBC driver version is less than 23.4.
+	 *
+	 * @return true if the driver version is less than 23.4, false otherwise
+	 */
+	protected boolean isVectorSupportedByDriver(WrapperOptions options) throws SQLException {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:91]: Instead of introducing a method that is repeatedly called, this is something that should be passed a boolean flag to the constructor of the various `JdbcType` classes.

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	@Override
+	public <X> ValueExtractor<X> getExtractor(final JavaType<X> javaTypeDescriptor) {
+
+		return new BasicExtractor<>( javaTypeDescriptor, this ) {
+			@Override
+			protected X doExtract(ResultSet rs, int paramIndex, WrapperOptions options) throws SQLException {
+				if ( isVectorSupportedByDriver(options) ) {
+					return javaTypeDescriptor.wrap( rs.getObject( paramIndex, javaType ), options );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:136]:
```suggestion
					return getJavaType().wrap( rs.getObject( paramIndex, ((AbstractOracleJdbcType) getJdbcType() ).getNativeJavaType() ), options );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+				}
+				else {
+					return javaTypeDescriptor.wrap( getVectorArray( rs.getString( paramIndex ) ), options );
+				}
+			}
+
+			@Override
+			protected X doExtract(CallableStatement statement, int index, WrapperOptions options) throws SQLException {
+				if ( isVectorSupportedByDriver(options) ) {
+					return javaTypeDescriptor.wrap( statement.getObject( index, javaType ), options );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:146]:
```suggestion
					return getJavaType().wrap( rs.getObject( index, ((AbstractOracleJdbcType) getJdbcType() ).getNativeJavaType() ), options );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+				else {
+					return javaTypeDescriptor.wrap( getVectorArray( statement.getString( index ) ), options );
+				}
+			}
+
+			@Override
+			protected X doExtract(CallableStatement statement, String name, WrapperOptions options)
+					throws SQLException {
+				if ( isVectorSupportedByDriver(options) ) {
+					return javaTypeDescriptor.wrap( statement.getObject( name, javaType ), options );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:157]:
```suggestion
					return getJavaType().wrap( rs.getObject( name, ((AbstractOracleJdbcType) getJdbcType() ).getNativeJavaType() ), options );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+				else {
+					return javaTypeDescriptor.wrap( getVectorArray( statement.getString( name ) ), options );
+				}
+			}
+
+
+		};
+	}
+
+	protected abstract <T> T getVectorArray(String string);
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:168]:
```suggestion
	protected abstract <T> T getVectorArray(String string);
	
	protected abstract Class<?> getNativeJavaType();
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+
+	protected abstract <T> String getStringVector(T vector, JavaType<T> javaTypeDescriptor, WrapperOptions options);
+
+	@Override
+	public int getDdlTypeCode() {
+		return getJdbcTypeCode();
+	}
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:175]:
```suggestion
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+				else {
+					st.setString( index, getStringVector( value, javaTypeDescriptor, options ) );
+				}
+			}
+
+			@Override
+			protected void doBind(CallableStatement st, X value, String name, WrapperOptions options)
+					throws SQLException {
+				if ( isVectorSupportedByDriver(options) ) {
+					st.setObject( name, value, OracleType.valueOf( getFriendlyName() ) );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:118]:
```suggestion
					st.setObject( name, value, getNativeTypeCode() );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+					return javaTypeDescriptor.wrap( getVectorArray( statement.getString( name ) ), options );
+				}
+			}
+
+
+		};
+	}
+
+	protected abstract <T> T getVectorArray(String string);
+
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:169]: Please add constants to `org.hibernate.dialect.OracleTypes`:

```suggestion

	protected abstract Class<?> getNativeTypeCode();
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+		appender.append( ", *, INT8)" );
+	}
+
+	@Override
+	public String getFriendlyName() {
+		return "VECTOR_INT8";
+	}
+
+	@Override
+	public int getJdbcTypeCode() {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java:46]:
```suggestion
	public int getDefaultSqlTypeCode() {
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleDoubleVectorJdbcType.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+		appender.append( ", *, FLOAT64)" );
+	}
+
+	@Override
+	public String getFriendlyName() {
+		return "VECTOR_FLOAT64";
+	}
+
+	@Override
+	public int getJdbcTypeCode() {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleDoubleVectorJdbcType.java:46]:
```suggestion
	public int getDefaultSqlTypeCode() {
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleFloatVectorJdbcType.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+		appender.append( ", *, FLOAT32)" );
+	}
+
+	@Override
+	public String getFriendlyName() {
+		return "VECTOR_FLOAT32";
+	}
+
+	@Override
+	public int getJdbcTypeCode() {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleFloatVectorJdbcType.java:47]:
```suggestion
	public int getDefaultSqlTypeCode() {
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorArgumentValidator.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+import org.hibernate.query.sqm.produce.function.FunctionArgumentException;
+import org.hibernate.query.sqm.tree.SqmTypedNode;
+import org.hibernate.type.BasicType;
+import org.hibernate.type.SqlTypes;
+import org.hibernate.type.spi.TypeConfiguration;
+
+/**
+ * A {@link ArgumentsValidator} that validates the arguments are all 4 oracle vector types i.e. {@link SqlTypes#VECTOR}, {@link SqlTypes#VECTOR_FLOAT32}, {@link SqlTypes#VECTOR_FLOAT64} and {@link SqlTypes#VECTOR_INT8}.
+ */
+public class OracleVectorArgumentValidator implements ArgumentsValidator {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorArgumentValidator.java:23]: I would prefer if you alter `VectorArgumentValidator` to support the additional type codes instead of adding a new validator.

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorJdbcType.java [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Specialized type mapping for generic vector {@link SqlTypes#VECTOR} SQL data type for Oracle.
+ * <p>
+ * This class handles generic vectors represented by an asterisk (*) in the format,
+ * allowing for different element types within the vector.
+ *
+ * @author Hassan AL Meftah
+ */
+public class OracleVectorJdbcType extends AbstractOracleVectorJdbcType {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorJdbcType.java:27]: This allows to inherit some methods and trim down this class.

```suggestion
public class OracleVectorJdbcType extends OracleFloatVectorJdbcType {
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+			final TypeConfiguration typeConfiguration = typeContributions.getTypeConfiguration();
+			final JavaTypeRegistry javaTypeRegistry = typeConfiguration.getJavaTypeRegistry();
+			final JdbcTypeRegistry jdbcTypeRegistry = typeConfiguration.getJdbcTypeRegistry();
+			final BasicTypeRegistry basicTypeRegistry = typeConfiguration.getBasicTypeRegistry();
+
+			populateVectorTypeCodes();
+			populateVectorBasicTypes();
+
+			// Ensure BasicTypes are resolved and stored for later use
+			registerSupportedBasicTypes( basicTypeRegistry );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java:74]: Resolving basic types must happen after jdbc types are registered.

File: hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+			}
+
+			@Override
+			protected X doExtract(CallableStatement statement, String name, WrapperOptions options)
+					throws SQLException {
+				if ( isVectorSupported ) {
+					return getJavaType().wrap( statement.getObject( name, ((AbstractOracleVectorJdbcType) getJdbcType() ).getNativeJavaType() ), options );
+				}
+				else {
+					return javaTypeDescriptor.wrap( getVectorArray( statement.getString( name ) ), options );
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/AbstractOracleVectorJdbcType.java:142]:
```suggestion
					return getJavaType().wrap( ( (AbstractOracleVectorJdbcType) getJdbcType() ).getVectorArray( statement.getString( name ) ), options );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+			populateVectorTypeCodes();
+			populateVectorBasicTypes();
+
+			boolean isVectorSupported;
+			try {
+				isVectorSupported = isVectorSupportedByDriver( (OracleDialect) dialect );
+			}
+			catch (SQLException e) {
+				isVectorSupported = false;
+			}
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java:80]:
```suggestion
			final boolean isVectorSupported = isVectorSupportedByDriver( (OracleDialect) dialect );
```

File: hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java [hunk 27]
```diff
@@ -0,0 +1,240 @@
...
+		vectorTypeCodes.put( float[].class, SqlTypes.VECTOR_FLOAT32 );
+		vectorTypeCodes.put( double[].class, SqlTypes.VECTOR_FLOAT64 );
+		vectorTypeCodes.put( byte[].class, SqlTypes.VECTOR_INT8 );
+	}
+
+	private void populateVectorBasicTypes() {
+		basicTypeReferences.put( float[].class, StandardBasicTypes.VECTOR_FLOAT32 );
+		basicTypeReferences.put( double[].class, StandardBasicTypes.VECTOR_FLOAT64 );
+		basicTypeReferences.put( byte[].class, StandardBasicTypes.VECTOR_INT8 );
+	}
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/OracleVectorTypeContributor.java:58]: Please remove

```suggestion
```

File: hibernate-vector/src/main/java/org/hibernate/vector/VectorArgumentValidator.java [hunk 28]
```diff
@@ -24,6 +25,13 @@ public class VectorArgumentValidator implements ArgumentsValidator {
 
 	public static final ArgumentsValidator INSTANCE = new VectorArgumentValidator();
 
+	private static final int[] availableVectorCodes = {
```
Comment (beikov) [inline hibernate-vector/src/main/java/org/hibernate/vector/VectorArgumentValidator.java:28]: Please remove

**Comment by loiclefevre** [comment]
🚀

**Comment by beikov** [comment]
Hello @Elab4d1,
I hope you’re well.

I'm going to need your help with a copyright matter: we’re exploring the possibility of changing the license of all Hibernate projects to Apache Software License v2 (ASL2).

You authored various contributions to our projects, some of which are currently licensed as LGPL2.1; I would love to have your explicit permission so that we might proceed with this change without further delays.

Could you reply and state that you agree to make all your contributions available under the ASL2?

The terms of the ASL2 can be found here: https://apache.org/licenses/LICENSE-2.0.

Also, if you have objections, please let me know. The Hibernate team believes that the ASL2 license would be more suitable for the future of our projects as it’s become more common and widely accepted; we think that in spirit and intent, it’s not a significant change, but would resolve some concerns that users have occasionally raised.

Please note that we’re not sure yet if we’ll be able to do this; I merely hope we can make it happen.

Regards,
Christian, on behalf of the Hibernate team

**Comment by Elab4d1** [comment]
Hello @beikov ,  

I confirm that I agree to make all my contributions to Hibernate ORM available under the Apache Software License 2.0.  

Best regards,  
Hassan
